### PR TITLE
Anchor support

### PR DESCRIPTION
**Pull request by robbielamb-wf**
## Ultimate problem:
Anchors need to be implemented for draft-2019 support.

## How it was fixed:
Do it.

## Testing suggestions:
Passing CI.

## Potential areas of regression:

---

> __FYA:__ @michaelcarter-wf

### CODE REVIEW

**Comment by aviary3-wk** [comment]
## Security Insights
No security relevant content was detected by automated scans.

## Action Items
* Review PR for [security impact](https://wiki.atl.workiva.net/display/SECURITY/Development+Security+Review+Guidelines); comment "security review required" if needed or unsure
* Verify [`aviary.yaml`](https://wiki.atl.workiva.net/pages/viewpage.action?pageId=42762752#SecurityInsightFramework&Tooling-ConfiguringAviary) coverage of security relevant code
---
_Questions or Comments? Reach out on Slack: #support-infosec._

File: lib/src/json_schema/json_schema.dart [hunk 1]
```diff
@@ -131,7 +131,7 @@ class JsonSchema {
           .future;
 
       // Boolean schemas are only supported in draft 6 and later.
-    } else if (data is bool && [SchemaVersion.draft6, SchemaVersion.draft7].contains(version)) {
+    } else if (data is bool && version != SchemaVersion.draft4) {
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:134]: How would you feel about using `version.compareTo(SchemaVersion.draft6) >= 0` for this?

File: lib/src/json_schema/json_schema.dart [hunk 2]
```diff
@@ -300,6 +298,18 @@ class JsonSchema {
           return _refMap[refPath] = rhs;
         }, mustBeValid: false);
       }
+    }
+
+    // Some sibling attributes depend on $id being setup first.
+    if (_schemaMap.containsKey(r'$id')) {
+      processAttribute(r'$id', _schemaMap[r'$id']);
+    }
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:306]: Nice approach!

File: lib/src/json_schema/json_schema.dart [hunk 3]
```diff
@@ -735,6 +746,9 @@ class JsonSchema {
   /// Base URI of the ID. All sub-schemas are resolved against this
   Uri _idBase;
 
+  /// Something like and ID, but different.
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:749]: We could probably be somewhat more helpful with this description :)

File: lib/src/json_schema/json_schema.dart [hunk 4]
```diff
@@ -1207,6 +1220,9 @@ class JsonSchema {
     return root.id;
   }
 
+  /// Anchor from something or other
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:1223]: lol

File: lib/src/json_schema/json_schema.dart [hunk 5]
```diff
@@ -1692,6 +1708,9 @@ class JsonSchema {
 
         // If the $id has a fragment, append it to the base, or use it alone.
       } else if (_id.fragment != null && _id.fragment.isNotEmpty) {
+        if (schemaVersion.hashCode >= SchemaVersion.draft2019_09.hashCode) {
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:1711]: Woah, why are we comparing hash codes like this?

File: lib/src/json_schema/json_schema.dart (hunk 3 above)
Comment (robbielamb-wf) [reply lib/src/json_schema/json_schema.dart:749]: oh man! this was supposed to be a temporary description....

File: lib/src/json_schema/json_schema.dart (hunk 5 above)
Comment (robbielamb-wf) [reply lib/src/json_schema/json_schema.dart:1711]: The hash code is just the version number. There should be a nicer way of doing it soon.

File: lib/src/json_schema/json_schema.dart (hunk 1 above)
Comment (robbielamb-wf) [reply lib/src/json_schema/json_schema.dart:134]: Talked in person. We'll have some followup work to clean up how we compare schema versions.

File: lib/src/json_schema/json_schema.dart (hunk 5 above)
Comment (tomdeering-wf) [reply lib/src/json_schema/json_schema.dart]: What an... amazing hashcode we have! /s

File: lib/src/json_schema/json_schema.dart [hunk 6]
```diff
@@ -1162,8 +1174,8 @@ class JsonSchema {
 
   /// The value of the exclusiveMaximum for the [JsonSchema], if any exists.
   num get exclusiveMaximum {
-    // If we're on draft6, the property contains the value, return it.
-    if ([SchemaVersion.draft6, SchemaVersion.draft7].contains(schemaVersion)) {
+    // If we're beyond draft4, the property contains the value, return it.
+    if (schemaVersion != SchemaVersion.draft4) {
```
Comment (tomdeering-wf) [inline lib/src/json_schema/json_schema.dart:1178]: `>` would probably be clearer than `!=`

Comment (tomdeering-wf) [reply lib/src/json_schema/json_schema.dart:1178]: Maybe we should change this to `schemaVersion >= SchemaVersion.draft6` for consistency.

**Review by tomdeering-wf — APPROVED** [review]

**Pull request merged**